Extend sdk to include setting conditions on Signing Coordinator

theref opened the pull request:
## Overview
This PR extends the TACo SDK to include functionality for setting conditions on the Signing Coordinator contract. It adds a new `setSigningCohortConditions` function that allows users to programmatically configure signing cohort conditions directly through the SDK, along with comprehensive unit and integration tests.

## Why
This provides a user-facing API for condition setting that was previously only available through direct contract interaction, making it easier for developers to manage signing cohort configurations.

## Implementation
Added the `setSigningCohortConditions` function to the sign module with proper parameter ordering (provider, domain, conditions, cohortId, chainId, signer), comprehensive unit tests covering success and error cases, and integration tests that validate condition serialization round-trips.

theref committed:
Add setSigningCohortConditions function

- Add setSigningCohortConditions method to SigningCoordinatorAgent
- Implement setSigningCohortConditions function in taco/sign.ts
- Export the new function from taco package index
- Function converts ConditionExpression to JSON and calls contract

theref committed:
Add comprehensive unit tests for setSigningCohortConditions

- Test successful condition setting with simple RPC conditions
- Test complex compound conditions (RPC + Contract)
- Test error handling from SigningCoordinatorAgent
- Use real condition objects instead of mocks for better validation
- Verify proper JSON serialization and contract method calls

theref committed:
Add condition serialization round-trip integration test

- Test validates condition serialization without blockchain writes
- Reads existing conditions from chain and converts to objects
- Converts objects back to JSON and verifies byte-perfect match
- Ensures read → object → write produces identical data
- Gracefully handles missing conditions on test cohorts

KPrasch reviewed (commented):
🚀 Finally a user-facing API for condition setting :-) looking forward to this one.

derekpierre commented on packages/shared/src/contracts/agents/signing-coordinator.ts:
```diff
@@ -57,6 +57,23 @@ export class SigningCoordinatorAgent {
+    domain: Domain,
+    cohortId: number,
+    chainId: number,
+    conditions: string,
```
Let's have the type for `conditions` be `Uint8Array` so that we keep the agent and the contract more closely aligned. The caller can take care of the conversion to bytes.

derekpierre commented on packages/taco/src/sign.ts:
```diff
@@ -173,3 +174,25 @@ export async function signUserOp(
+  domain: Domain,
+  cohortId: number,
+  chainId: number,
+  signer: ethers.Signer,
```
A couple things:
1. To remain consistent with `encrypt`, let's take a `Condition` here, internally we can convert it to a `ConditionExpression`.
2. The existing taco APIs (`encrypt`, `decrypt`, `signUserOp`) have an implicit order of parameters that we should try to remain consistent with. `provider` first, `domain` second, ..., `signer` last.

```suggestion
  provider: ethers.providers.JsonRpcProvider,
  domain: Domain,
  conditions: Condition,
  cohortId: number,
  chainId: number,
  signer: ethers.Signer,
```

wdyt?

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
       expect(signResult.signingResults).toBeDefined();
       expect(Object.keys(signResult.signingResults).length).toBeGreaterThan(0);
-    }, 15000);
+    }, 150000);
```
This was fixed before, perhaps undone as part of your rebase?

In any case, it should be 15s and not 150s.

```suggestion
    }, 15000);
```

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+          console.log(`Skipping test - no conditions found for cohort ${COHORT_ID}. This test requires an existing cohort with conditions set.`);
+          return;
+        }
+        throw error;
```
I don't believe this is necessary. The contract returns empty bytes for a non-existent cohort ID or non-set conditions for a chain for a specific cohort Id. So you probably just need to check if the bytes are empty.

Also, just noticed that the definition of `getSigningCohortConditions` for the SigningCoordinatorAgent returns `Promise<string>`. Is that because it returns a hex string instead of bytes?

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+        return;
+      }
+      
+      console.log('Retrieved condition hex from chain:', retrievedConditionHex);
```
Let's skip on the printing - the terminal gets flooded with these (I run integration tests locally 😄 ). Applies to other locations below.

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+      if (!retrievedConditionHex || retrievedConditionHex === '0x' || retrievedConditionHex.length <= 2) {
+        console.log(`Skipping test - cohort ${COHORT_ID} has no conditions set.`);
+        return;
+      }
```
I think do an `expect` instead to check that the conditions are set - if not set, then the test will fail and we'll update the values used. It's a way of explicitly making sure that the test is being run and if it fails because the cohortID/chainID combination no longer has a condition we should update the values or extend the ritual etc.

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+      
+      console.log('Condition serialization round-trip validation completed successfully!');
+      
+    }, 60000); // Shorter timeout since no blockchain writes
```
I think you meant 6s instead of 60s.
```suggestion
    }, 6000); // Shorter timeout since no blockchain writes
```

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+        DOMAIN as Domain,
+        COHORT_ID,
+        CHAIN_ID,
+      );
```
Maybe just do an `expect(...).not.toThrow()` since `retrievedContext` isn't used after this line?

derekpierre commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -98,6 +100,76 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
+      const roundTripJson = retrievedConditions.toJson();
+      console.log('Round-trip JSON:', roundTripJson);
+      
+      // The core verification: ensure serialization round-trip is identical
+      // This proves that read → object → write produces identical bytes
+      expect(JSON.parse(roundTripJson)).toEqual(JSON.parse(chainConditionJson));
```
This check is repeated on L169?

derekpierre reviewed (commented):
Seems like netlify can't launch demos - let's look into why that is as well.

theref committed:
Add .claude/ directory to .gitignore

theref committed:
Refactor setSigningCohortConditions parameter order

Move provider and domain parameters first for consistency with other agent methods.
Change conditions parameter type from ConditionExpression to Condition for improved API ergonomics.

theref committed:
Update unit tests for new setSigningCohortConditions signature

Update test calls to match new parameter order and use Condition objects instead of ConditionExpression.

theref committed:
Improve integration test reliability and clarity

- Remove debug console.log statements
- Reduce test timeouts to more reasonable values (15s, 6s)
- Simplify condition serialization test logic by removing unnecessary error handling
- Add proper assertions for condition validation

theref committed:
Change SigningCoordinatorAgent conditions parameter to Uint8Array

Keep agent interface closer to contract by accepting Uint8Array instead of string.
Remove internal UTF-8 conversion to let caller handle the conversion.

theref committed:
Change setSigningCohortConditions to accept Uint8Array parameter

Move condition-to-bytes conversion responsibility to caller as suggested by @derekpierre.
Update function signature to take Uint8Array directly instead of Condition.
Update unit tests to handle conversion from Condition to Uint8Array.

Co-authored-by: Derek Pierre <derek@nucypher.com>

theref committed:
Remove unused imports from sign.ts

Remove unused Condition and ConditionExpression imports after API changes.

derekpierre commented on packages/taco/src/sign.ts:
```diff
@@ -173,3 +173,22 @@ export async function signUserOp(
+export async function setSigningCohortConditions(
+  provider: ethers.providers.JsonRpcProvider,
+  domain: Domain,
+  conditions: Uint8Array,
```
Apologies, I likely didn't explain the overall picture well.

This external facing API should use `Condition` similar to `encrypt`/`encryptWithPublicKey`. Subsequently (within this function) we can create the `ConditionExpression` and then convert it to bytes before calling `setSigningCohortConditions` on the agent.

theref replied:
ahhhh ok, I was quite surprised! I should've pushed back a bit, sorry

derekpierre replied:
Definitely push back if you ever disagree and/or think that I sound off my rocker 😅 

The underlying philosophy, at least for me, boils down to:
- Agent function parameters should pretty much match the contract call parameters - not much, if any, additional processing should be done in agents.
- SDK/API parameters should be simple / consistent for the external caller of the API - not too much processing for them to do before calling
- Our implementation of the API that connects the two (API and agents), can do the additional processing

derekpierre commented on .gitignore:
```diff
@@ -16,3 +16,4 @@ docs-json
 .next
 packages/docs
 docs
+.claude/
```
Hehe interesting.

theref replied:
yeah I'm experimenting a lot at the moment. I'm pen to hear what other people are doing and whether we want to standardise some stuff

theref committed:
Implement correct API design for setSigningCohortConditions

Fix misunderstanding of @derekpierre's feedback by implementing two-layer design:
- Public API: Takes user-friendly Condition parameter (like encrypt())
- Agent layer: Takes Uint8Array for close contract alignment
- Internal conversion: Public API handles Condition → ConditionExpression → JSON → Uint8Array

This provides optimal developer experience while maintaining efficient internal representation.

Co-authored-by: Derek Pierre <derek@nucypher.com>

derekpierre commented:
Seems like the netlify demo previews are failing (https://github.com/nucypher/taco-web/pull/681#issuecomment-3061366084, https://github.com/nucypher/taco-web/pull/681#issuecomment-3061369788). From the looks of the log, this seems to be the issue:

```
5:05:53 AM: Failed during stage 'Install dependencies': dependency_installation script returned non-zero exit code: 1
5:05:53 AM:  ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with <ROOT>/package.json
5:05:53 AM: Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"
5:05:53 AM:   Failure reason:
5:05:53 AM:   specifiers in the lockfile don't match specifiers in package.json:
* 2 dependencies are mismatched:
5:05:53 AM:   - @vitest/coverage-v8 (lockfile: ^3.0.9, manifest: ^3.2.4)
5:05:53 AM:   - eslint-plugin-unused-imports (lockfile: ^3.2.0, manifest: ^4.1.4)
```

It was probably some mismatch that perhaps happened during a rebase. I think you can run `pnpm install --fix-lockfile` from the root folder to resolve the problem.

derekpierre commented on packages/taco/src/sign.ts:
```diff
@@ -173,3 +175,27 @@ export async function signUserOp(
+  signer: ethers.Signer,
+): Promise<ethers.ContractTransaction> {
+  // Convert Condition to ConditionExpression, then to JSON, then to bytes
+  const conditionExpression = new ConditionExpression(conditions);
+  const conditionsJson = conditionExpression.toJson();
+  const conditionsBytes = ethers.utils.toUtf8Bytes(conditionsJson);
```
🔥

derekpierre reviewed (commented):
Once we get the netlify previews working, I'll approve.

theref committed:
Fix workspace dependency resolution for demos

Regenerate lockfile to ensure demos use workspace packages, fixing Netlify deployment issues.

derekpierre reviewed (approved):
🎸 .

The netlify issue seems separate, and hasn't run on a new commit.
